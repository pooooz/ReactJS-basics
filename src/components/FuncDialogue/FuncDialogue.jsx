import React, {useState, useEffect, useRef} from "react";
import styles from './FuncDialogue.module.scss'
import {FuncMessageList} from "./components/FuncMessageList/FuncMessageList";
import {Input} from "./components/Input/Input";

export const FuncDialogue = () => {
  const [messages, setMessages] = useState([{text: 'Initialization...', author: 'Admin'}]);
  const [inputValue, setInputValue] = useState('');
  const [botMessageCounter, setBotMessageCounter] = useState(0);

  const handleChange = (event) => {
    setInputValue(event.target.value);
  }

  const handleSubmit = (event) => {
    if (inputValue !== '') {
      setMessages([...messages, {text: inputValue, author: 'You'}]);
      setInputValue('');
      setBotMessageCounter(botMessageCounter + 1);
    }
    event.preventDefault();
  }

  const botMessageCounterRef = useRef(botMessageCounter);
  useEffect(() => {
    if (botMessageCounterRef.current !== botMessageCounter) {
      setTimeout(() => {
        setMessages([...messages, {text: '[BOT] Message ' + botMessageCounter, author: 'BOT'}]);
      }, 1500);
    }
  }, [botMessageCounter]);

  return (
    <section className={styles.dialogue}>
      <FuncMessageList messages={messages}/>
      <form className={styles.form} onSubmit={handleSubmit}>
        <Input value={inputValue} change={handleChange}/>
        <button className={styles.form__button} type='submit'>Send</button>
      </form>
    </section>
  );
}