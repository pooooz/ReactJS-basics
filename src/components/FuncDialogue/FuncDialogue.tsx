import React, { useState, useEffect, useRef, FC } from 'react';
import styles from './FuncDialogue.module.scss';
import { FuncMessageList } from './components/FuncMessageList/FuncMessageList';
import { Input } from './components/Input/Input';
import { Button, ThemeProvider } from '@mui/material';
import { createTheme } from '@mui/material/styles';

const theme = createTheme({
  typography: {
    fontFamily: 'Press Start 2P',
  },
});

export interface Message {
  text: string;
  author: string;
}

const FuncMessageListPure = React.memo(FuncMessageList);

export const FuncDialogue: FC = () => {
  const [messages, setMessages] = useState<Message[]>([
    { text: 'Initialization...', author: 'Admin' },
  ]);
  const [inputValue, setInputValue] = useState<string>('');
  const [botMessageCounter, setBotMessageCounter] = useState<number>(0);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(event.target.value);
  };

  const handleSubmit = (event: React.FormEvent) => {
    if (inputValue) {
      setMessages([...messages, { text: inputValue, author: 'You' }]);
      setInputValue('');
      setBotMessageCounter(botMessageCounter + 1);
    }
    event.preventDefault();
  };

  const botMessageCounterRef = useRef(botMessageCounter);
  useEffect(() => {
    if (
      messages[messages.length - 1].author !== 'BOT' &&
      botMessageCounterRef.current !== botMessageCounter
    ) {
      const timeout = setTimeout(() => {
        setMessages([...messages, { text: '[BOT] Message', author: 'BOT' }]);
      }, 1500);
      return () => clearTimeout(timeout);
    }
  }, [botMessageCounter, messages]);

  return (
    <ThemeProvider theme={theme}>
      <section className={styles.dialogue}>
        <FuncMessageListPure messages={messages} />
        <form className={styles.form} onSubmit={handleSubmit}>
          <Input value={inputValue} change={handleChange} />
          <Button
            className={styles.form__button}
            type="submit"
            style={{
              width: '100px',
              fontFamily: 'Press Start 2P',
              color: '#32CD32FF',
            }}
          >
            Send
          </Button>
        </form>
      </section>
    </ThemeProvider>
  );
};
