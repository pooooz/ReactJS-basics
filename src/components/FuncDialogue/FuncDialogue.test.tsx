import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import '@testing-library/user-event';
import { FuncDialogue } from './FuncDialogue';

describe('FuncDialogue', () => {
  it('Render component', () => {
    render(<FuncDialogue />);
  });

  it('Render with snapshot', () => {
    const { asFragment } = render(<FuncDialogue />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Input test', () => {
    render(<FuncDialogue />);
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: 'Test' },
    });
    expect(screen.getByRole('textbox')).toContainHTML('Test');
  });

  it('Forming a message', () => {
    render(<FuncDialogue />);
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: 'Test' },
    });
    fireEvent.click(screen.getByRole('button'));
    expect(screen.getByText(/<You>/)).toBeInTheDocument();
  });

  it('Forming a empty message', () => {
    const { asFragment } = render(<FuncDialogue />);
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: '' },
    });
    fireEvent.click(screen.getByRole('button'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('Bot answer', async () => {
    render(<FuncDialogue />);
    fireEvent.input(screen.getByRole('textbox'), {
      target: { value: 'Test' },
    });
    fireEvent.click(screen.getByRole('button'));
    await waitFor(
      () => {
        expect(screen.getByText(/\[BOT\] Message/)).toBeInTheDocument();
      },
      { timeout: 1600 }
    );
  });
});
