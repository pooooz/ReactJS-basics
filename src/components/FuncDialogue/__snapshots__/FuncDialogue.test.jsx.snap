// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FuncDialogue Forming a empty message 1`] = `
<DocumentFragment>
  <section>
    <div>
      <div>
        <p
          data-text="Initialization..."
        >
          Initialization...
        </p>
        <p
          data-text="Author: <Admin>"
        >
          Author: &lt;Admin&gt;
        </p>
      </div>
    </div>
    <form>
      <div>
        <input
          type="text"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Send
      </button>
    </form>
  </section>
</DocumentFragment>
`;

exports[`FuncDialogue Render with snapshot 1`] = `
<DocumentFragment>
  <section>
    <div>
      <div>
        <p
          data-text="Initialization..."
        >
          Initialization...
        </p>
        <p
          data-text="Author: <Admin>"
        >
          Author: &lt;Admin&gt;
        </p>
      </div>
    </div>
    <form>
      <div>
        <input
          type="text"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Send
      </button>
    </form>
  </section>
</DocumentFragment>
`;
